### PR TITLE
C++ Fold nn module

### PR DESCRIPTION
**Pull request by ShahriarRezghi**

### CODE REVIEW

**Comment by ShahriarSS** [comment]
Why does this PR have ```caffe2``` label?

File: test/cpp/api/modules.cpp [hunk 1]
```diff
@@ -118,6 +119,23 @@ TEST_F(ModulesTest, Linear) {
...
+
+  s.backward();
+  ASSERT_EQ(y.ndimension(), 4);
+  ASSERT_EQ(s.ndimension(), 0);
+  ASSERT_EQ(y.size(0), 1);
+  ASSERT_EQ(y.size(1), 3);
+  ASSERT_EQ(y.size(2), 4);
+  ASSERT_EQ(y.size(3), 5);
+
+  // TODO check numel of grad
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: Do we plan to implement this part as well?

File: torch/csrc/api/include/torch/nn/modules/fold.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  /// number of points for each dimension before reshaping.
+  TORCH_ARG(ExpandingArray<2>, padding) = 0;
+
+  /// controls the stride for the sliding blocks.
+  TORCH_ARG(ExpandingArray<2>, stride) = 1;
+};
+
+/// Applies fold over a 3-D input.
+/// See https://pytorch.org/docs/master/nn.html#torch.nn.Fold to learn about
+/// the exact behavior of this module.
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/fold.h]: The documentation here looks great. We can experiment with porting all docs (including math formulas) from https://pytorch.org/docs/stable/nn.html#fold to here in this PR, or in a later PR if you prefer.

**Review by yf225 — COMMENTED** [review]
Thanks a lot for the PR @ShahriarSS! Overall it looks awesome. Once https://github.com/pytorch/pytorch/pull/23852 lands we can add the `Fold` module into the API parity whitelist, and then we can merge this PR.

**Comment by yf225** [comment]
@ShahriarSS Regarding the `caffe2` label: it's probably because we changed the CMakeLists.txt file within the `caffe2/` folder. This change is not really Caffe2 and we can ignore the label.

File: test/cpp/api/modules.cpp (hunk 1 above)
Comment (ShahriarSS) [reply test/cpp/api/modules.cpp]: Yes. I just forgot about it.

File: torch/csrc/api/include/torch/nn/modules/fold.h (hunk 2 above)
Comment (ShahriarSS) [reply torch/csrc/api/include/torch/nn/modules/fold.h]: I'll add the formulas here. We also need to do this for the other modules already implemented.

**Comment by ShahriarSS** [comment]
@yf225 I removed the TODO from fold test because fold doesn't have weights of any kind. I think ```nn::Fold``` is ready.

**Comment by yf225** [comment]
@ShahriarSS Thanks! I am working on adding API parity test for `nn::Fold` (currently blocked by https://github.com/pytorch/pytorch/pull/25784), and I will merge the changes into this PR and then land it.

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -513,6 +513,7 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/conv.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/dropout.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/embedding.cpp
+      ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/fold.cpp
```
Comment (yf225) [inline caffe2/CMakeLists.txt]: @ShahriarSS the Facebook internal build seems to fail, and I am thinking we might need to add the `fold.cpp` entry into https://github.com/pytorch/pytorch/blob/13292ec3c7f47228b93dbef1ca750b66b4b8e0af/tools/build_variables.py#L178-L205
